src/app/modules/sgid/__tests__/sgid.service.spec.ts

```diff
@@ -0,0 +1,234 @@
+import { SgidClient } from '@opengovsg/sgid-client'
+import fs from 'fs'
+import { mocked } from 'ts-jest/utils'
+
+import {
+  SgidCreateRedirectUrlError,
+  SgidFetchAccessTokenError,
+  SgidFetchUserInfoError,
+  SgidInvalidJwtError,
+  SgidInvalidStateError,
+  SgidVerifyJwtError,
+} from '../sgid.errors'
+import { SgidService } from '../sgid.service'
+
+import {
+  MOCK_ACCESS_TOKEN,
+  MOCK_AUTH_CODE,
+  MOCK_DESTINATION,
+  MOCK_JWT,
+  MOCK_JWT_PAYLOAD,
+  MOCK_NONCE,
+  MOCK_OPTIONS,
+  MOCK_REDIRECT_URL,
+  MOCK_REMEMBER_ME,
+  MOCK_STATE,
+  MOCK_TOKEN_RESULT,
+  MOCK_USER_INFO,
+} from './sgid.test.constants'
+
+jest.mock('@opengovsg/sgid-client')
+const MockSgidClient = mocked(SgidClient, true)
+jest.mock('fs', () => ({
+  ...(jest.requireActual('fs') as typeof fs),
+  readFileSync: jest.fn().mockImplementation((v) => v),
+}))
+
+describe('sgid.service', () => {
+  beforeEach(async () => {
+    jest.clearAllMocks()
+  })
+  describe('constructor', () => {
+    it('should create an SgidClient correctly', () => {
+      const { endpoint, clientId, clientSecret, privateKey, redirectUri } =
+        MOCK_OPTIONS
+      const sgidService = new SgidService(MOCK_OPTIONS)
+      expect(sgidService).toBeInstanceOf(SgidService)
+      expect(MockSgidClient).toHaveBeenCalledWith({
+        endpoint,
+        clientId,
+        clientSecret,
+        privateKey,
+        redirectUri,
+      })
+    })
+  })
+  describe('createRedirectUrl', () => {
+    it('should return a string if ok', () => {
+      const sgidService = new SgidService(MOCK_OPTIONS)
+      const sgidClient = mocked(MockSgidClient.mock.instances[0], true)
+      sgidClient.authorizationUrl.mockReturnValue({
+        url: MOCK_REDIRECT_URL,
+        nonce: MOCK_NONCE,
+      })
+      const url = sgidService.createRedirectUrl(MOCK_STATE)
+      expect(url._unsafeUnwrap()).toEqual(MOCK_REDIRECT_URL)
+      expect(sgidClient.authorizationUrl).toHaveBeenCalledWith(MOCK_STATE)
+    })
+    it('should return error if not ok', () => {
+      const sgidService = new SgidService(MOCK_OPTIONS)
+      const sgidClient = mocked(MockSgidClient.mock.instances[0], true)
+      sgidClient.authorizationUrl.mockReturnValue({
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        url: undefined,
```

Comment:
🤔

Reply:
I need this there to induce SgidService to throw up